### PR TITLE
Implement SPI mode

### PR DESCRIPTION
**Pull request by hasenbanck**
Implements flashing over SPI. Tested on a STM32F722ZE. Speed improvement over flashing with I2C is around 24x.

Changes:
 * Implements SPI backend
 * Improves flashing logic to handle edge cases of SPI protocoll
 * Timeout improvements
 * Added help / man information
 * Added Author
 * Increased version number

### CODE REVIEW

**Comment by fpistm** [comment]
Hi @hasenbanck,

thanks for this feature.
I wonder if it's should not be good to propose it to https://sourceforge.net/projects/stm32flash even if it seems not to have lot of activities.
I've "forked" this repo to be able to propose the release for all host OS (used by the board manager ) and also be able to propose any patch quickly if required.
Anyway I can add to stm32duino version and if it is merged in the official repo then I will sync.

**Comment by hasenbanck** [comment]
@fpistm Yeah, I think I will also try that. But it really seems dead, since there are three PR of long time contributors that are sitting there for over one year. But I will give it a shot.

If the project is really dead, then maybe it's time for a fork, so that the floating changes can finally reach a new version, since the version that most distributions ship with is rather old and is missing some MCU support.

**Comment by fpistm** [comment]
Right.
I will ask Antonio to have a look on your PR. ;)

File: spi.c [hunk 1]
```diff
@@ -0,0 +1,272 @@
+/*
+  stm32flash - Open Source ST STM32 flash program for *nix
+  Copyright (C) 2014 Antonio Borneo <borneo.antonio@gmail.com>
```
Comment (fpistm) [inline spi.c]: I guess you start this file from a copy of the I2C ?
Then I think you could add your copyright also as it is done in other .c file:
https://github.com/stm32duino/stm32flash/blob/d8c265f4b17861213cfb183a8fb656291215da97/init.c#L3
```
  stm32flash - Open Source ST STM32 flash program for *nix
  Copyright (C) 2010 Geoffrey McRae <geoff@spacevs.com>
  Copyright (C) 2013 Antonio Borneo <borneo.antonio@gmail.com>

```

**Comment by hasenbanck** [comment]
Testing my implemenation, I think I broke the UART interface. I will hunt that bug in the comming days.

**Comment by hasenbanck** [comment]
And I can't reproduce the problem now. Maybe it was a bad connection? I will keep an eye on it and will test it more.

**Comment by hasenbanck** [comment]
@fpistm No need to hurry. It's just a feature nice to have for my Heteromycin board, since you could flash the board over a web interface pretty fast this way.

I just pushed a commit that cleans up the code. It removes some workaround I had to include to better handle edge cases in a fast way and this commit now implements these edgecases in a way without the need of workarounds.

I will shamelessly ping somebody who might be interested in this PR. 

@ktand Feel free to ignore this if you are not interested.

**Comment by hasenbanck** [comment]
Some issues I found in general with my F765VI and the current STM32flash code:
- .BIN files seem to flash wrong (compiled by arduino). But .hex files are parsed and flashed alright
- I2C seems to run into a problem when trying to mass erase a STM32F765VI. STM32F4 works fine. STM32F7 I2C uses procotol v1.2. whereas ST32F4 seem to use v1.1.

Problems with this branch and F765:
- Flash over UART fails at mass erase stage
- I2C has the same mass erase problem
- SPI erases and starts flashing, but stops most of the time or the flash afterwards is corrupt.

My tests with the F4 on the other hand are all fine in these particular cases. I will try to debug this particular combination more and try to fix all bugs one after the other.

**Comment by fpistm** [comment]
Hi @hasenbanck,

I close this PR and will put the repo on hold.
The STM32 core does not use anymore this tool since STM32CubeProgrammer has been deployed.

**Pull request closed without merging**